gm8emulator/src/game/audio.rs

```diff
@@ -44,41 +48,148 @@ pub struct SoundParams {
+        let (sample_sender, sample_reciever) = mpsc::channel();
 
-        std::thread::spawn(move || {
-            let stream = session.open_output_stream(device).unwrap();
-            stream.play(mixer).unwrap();
+        let interprocess_source = InterprocessSource::new(sample_reciever, channel_count, sample_rate);
+
+        let audio_dumper = do_dump_audio.then(|| {
+            process::Command::new("ffmpeg")
+                .arg("-y")
+                .arg("-f")
+                .arg("f32le")
+                .arg("-ar")
+                .arg(sample_rate.to_string())
+                .arg("-ac")
+                .arg(channel_count.to_string())
+                .arg("-i")
+                .arg("-")
+                .arg("dump.flac")
+                .stdin(Stdio::piped())
+                .stdout(Stdio::null())
+                .stderr(Stdio::null())
+                .spawn()
+                .expect("Failed to open FFmpeg stdin")
         });
 
-        Self {
-            mixer_handle,
-            mixer_channel_count: channel_count,
-            mixer_sample_rate: sample_rate,
-            do_output,
-            global_volume,
-            end_times: HashMap::new(),
-            multimedia_end: None,
+        if do_dump_audio {
+            std::thread::spawn(move || {
+                let stream = session.open_output_stream(device).unwrap();
+
+                stream.play(interprocess_source).unwrap();
+            });
+
+            return Self {
+                mixer: Some(mixer),
+                sample_sender: Some(sample_sender),
+                mixer_handle,
+                mixer_channel_count: channel_count,
+                mixer_sample_rate: sample_rate,
+                do_output,
+                global_volume,
+                end_times: HashMap::new(),
+                multimedia_end: None,
+                audio_dumper: audio_dumper,
+            };
+        } else {
+            std::thread::spawn(move || {
+                let stream = session.open_output_stream(device).unwrap();
+
+                stream.play(mixer).unwrap();
+            });
+
+            Self {
+                mixer: None,
+                sample_sender: None,
+                mixer_handle,
+                mixer_channel_count: channel_count,
+                mixer_sample_rate: sample_rate,
+                do_output,
+                global_volume,
+                end_times: HashMap::new(),
+                multimedia_end: None,
+                audio_dumper: audio_dumper,
+            }
         }
     }
+    pub fn dump_audio(&mut self) {
+        if let Some(mixer) = &mut self.mixer {
+            const AUDIO_SAMPLES_PER_FRAME: usize = 48000 / 50 * 2;
+
+            let mut audio_output: [Sample; AUDIO_SAMPLES_PER_FRAME] = [0.0; AUDIO_SAMPLES_PER_FRAME];
+
+            mixer.write_samples(&mut audio_output);
+
+            let stdin = self.audio_dumper.as_mut().unwrap().stdin.as_mut().expect("Failed to open stdin");
```

Comment:
I believe writing samples shall take place after opening `stdin`, so as not to do work in vain if it fails.
```suggestion
            let stdin = self.audio_dumper.as_mut().unwrap().stdin.as_mut().expect("Failed to open stdin");
            mixer.write_samples(&mut audio_output);
```